### ISSUE TITLE
Removing section cut object does not remove node group

### ISSUE THREAD

**Issue by schroef** [IfcOpenShell/IfcOpenShell#2168]
v0.09999 ?!?!?!)
bl2.93 Win

I was testing out this addon after looking for the cutawayshader. I noticed a link in the blendercommunity post and was happy to see this one works with Eevee as the other one is OSL and thus CPU only.

I was testing the addon only and noticed when removing the section cut object, the nodegroup is not removed from the materials.

tried it couple times and each time nodegroup stays in the scene. Is this perhaps im testing it outside of the BlenderBIM addon?

**Comment by Moult** [comment]
v0.0.999999 means you are using an unstable release which is unversioned :) The 99999 convention comes from Gentoo package management.

Indeed the nodegroup isn't removed it seems. Ping @Gorgious56 who helped implement this, perhaps he knows a quick fix?

**Comment by Gorgious56** [comment]
Indeed there is AFAICR no callback to ensure the shader is removed when the cutaway object is removed.
Perhaps something can be implemented using the msgbus system, or simply adding a button to clear the cutaway shader from the material.
IIRC you should first remove the section cutaway from the interface, and then you can delete the section cutaway object

Are you using the standalone script (https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/blenderbim/scripts/standalone_section_shader.py) or the feature inside the BlenderBIM addon ?

**Comment by Moult** [comment]
Is it too onerous to loop through every material in the project and check that way? A little brute force but it only ever needs to run when there are no more section compare nodes in the node group?

**Comment by Gorgious56** [comment]
I think the problem here is that when the object is removed by the user using X (delete) and not using the button in the addon panel, the addon is never notified of it.

I don't think it would be computationally expensive to parse every material that implements the cutaway shader.
Removing the node group from a material if there is no more section compare inside it can be implemented : when removing a section compare, check if this was the last one, run through every material, disconnect the node group where it makes sense. Although it won't fix the above problem.

There is also this problem that I'm sure you're aware of that an object can be modified / moved / deleted in a myriad of different ways in Blender, so it may not be possible to account for every one of them. For example deleting an object in the 3D viewport is not the same thing as deleting an object from the outliner.

**Comment by Moult** [comment]
I think that's OK, two options I can think of:

 1. Let the user remove the section object, but whenever the user next presses add / remove section, a "clean up" is done in the node tree.
 2. We actually already hook into object deletion via a few methods, so in theory it could be added there. https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/blenderbim/blenderbim/bim/module/geometry/operator.py#L322

**Comment by schroef** [comment]
> v0.0.999999 means you are using an unstable release which is unversioned :) The 99999 convention comes from Gentoo package management.
> 
> Indeed the nodegroup isn't removed it seems. Ping @Gorgious56 who helped implement this, perhaps he knows a quick fix?

This is what is the master branch i found. See src > blenderbim >  scripts > [standalone_section_shader.py](https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/blenderbim/scripts/standalone_section_shader.py)

i was checking on my old osx system, there it does remove the texture coordinates and "section compare" group. Weird thing is. If i manually disconnect it, the meshes still render fine. If the delete button is used, nothing shows. Thats super weird. It also causes my old blender (2.83) to crash. When i disconnect the group output and reconnect it, than it does show again in the render. Feels like it needs some refresh. 

Another issue is the delete button has a poll for active object, but thats wrong. When something different is selected that delete button looks like it can be used.

Ive added a crash report from blender on OSX, not sure it helps. I find these kinda complicated to look at.
[Test scene-01.crash.txt](https://github.com/IfcOpenShell/IfcOpenShell/files/8584592/Test.scene-01.crash.txt)

**Comment by schroef** [comment]
> from

Perhaps this is a solution for my issue i see on OSX. Stated in my prior message, on OSX using bl2.83 it does delete, but then it render most times blank or it crashes blender. The weird thing is when i manually disconnect that compare part from the last mix shader node, it does work and the models show in the render. PS i noticed this issue while doing a render and then delete the section object.

It looks like when i add a refresh in one of the nodes, it renders fine. So in this case i set the factor of the mix node inside "Section Override" to 1 again. Im using this command ```section_override.nodes["Section Mix"].inputs[0].default_value = 1```

**Comment by schroef** [comment]
Ive also added a new line to the poll operator for deleting section object. This way the button will only highlight when section objects are selected. Though perhaps it should also have check to see if its and empty
``@classmethod
    def poll(cls, context):
        aob = context.active_object
        return ("Section" in aob.name) and bpy.data.node_groups.get("Section Override") and aob is not None```

**Comment by schroef** [comment]
> Is it too onerous to loop through every material in the project and check that way? A little brute force but it only ever needs to run when there are no more section compare nodes in the node group?

Looping over all materials is only needed when a user has used the X to delete section object. The nodegroup is used in away that every material gets the exact same node group. Nothing is exposed of the nodegroup and therefor all materials have the same.

Perhaps a handler could be used for post render. This would check all items before starting a render and clean it. But this can be done easier perhaps. Why not let it check if there are empties named "section" and then compare that to the amount of users for the "Section Compare" nodegroup. If this mismatches, then we now it was manually deleted. Though i guess the issue is the texture coordinates is pointing to the object which is named "Section.XXX"

**Comment by Gorgious56** [comment]
Hello, you may try out the script I modified in [this PR](https://github.com/IfcOpenShell/IfcOpenShell/pull/2172) and see if it improves your problems. It should be less error prone overall.

About the poll method you suggested, I'm always hesitant to check against object names because the user or other parts of the script can modify it. In the end I only kept the test against the node group actually existing because I extended the operator to delete any selected section object. See https://github.com/IfcOpenShell/IfcOpenShell/blob/c40f90011ee631f62dd229a2b7be6b199faf8fa9/src/blenderbim/scripts/standalone_section_shader.py#L322-L323

Of course under the hood it checks the name of the node group, but it's harder for the layman to inadvertently modify the name of a node group than the name of an object.

The cleaning behaviour checks every texture coordinate node to see if there is an object associated with it, if there is none delete it, if there is one and it has only one user (the texture coordinate node) delete it, otherwise check the next one. If there is no texture coordinate node in the tree, scrap everything clean.

I'm open to any criticisism and feedback of course.

**Comment by schroef** [comment]
I also see an issue with the option "Add to selected objects", that function is only checked when an object doesnt have any materials. The result currently is that everything in the scene will get the section cut applied.

**Comment by schroef** [comment]
PS that new version look much nicer! Like the panel layout. I had done about the same in my adjust version. Yet i see you have totally overhauled everything, adding into a main class is much cleaner. Im not that much familiar with how to work with that. What is difference of adding all those static functions inside a class vs using those function outside any class?
![Screen Shot 2022-04-28 at 18 16 07](https://user-images.githubusercontent.com/6923008/165856716-b191a5a8-9168-41b1-bbb4-1e01ce327080.png)


I think i got my version now work correctly when using "Only Selected".

**Comment by Moult** [comment]
@schroef you may be interested in reading https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/blenderbim/docs/blenderbim/developer_guide.rst which talks a bit about the architecture of the BlenderBIM Add-on. In short, it's a very big add-on and Blender operators, abstract domain logic, and the concrete implementation of that logic are all separated for fast and comprehensive testing, and code reuse.

It has less relevance in the standalone version :)

**Comment by schroef** [comment]
> @schroef you may be interested in reading https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.7.0/src/blenderbim/docs/blenderbim/developer_guide.rst which talks a bit about the architecture of the BlenderBIM Add-on. In short, it's a very big add-on and Blender operators, abstract domain logic, and the concrete implementation of that logic are all separated for fast and comprehensive testing, and code reuse.
> 
> It has less relevance in the standalone version :)

I read a bit about it this week, looks very impressive! I was not that interested in that now, i was mainly interested in this cutaway or section plane shader

**Comment by Gorgious56** [comment]
> 
> EDIT im not quite there. I still see an issue where when i select another object it also gets the other section cut object. I believe this is because other operators running first and not checking for possible other nodegroups using the same name. I see its always referencing to "Section Override". I think if we do the check if it needs to be used for a single object we need to store that nodegroup name, return it and then use that same name in the other operators.

For now there is only one section cutaway system so all cutter objects are shared in one underlying node group. In the future the system can be extended to support several different section cutaway systems, but we'd need a comprehensive UI so that it's easy to use and tweak.

If you're okay with it, don't hesitate to share your code so I can see if I can incorporate your changes into the pull request and into the addon codebase.

Cheers

**Comment by Moult** [comment]
By the way don't let me hold you back happy for you to merge the PR anytime :)

**Comment by schroef** [comment]
Found the new versions today and was glad to see it can also work on collections now. Was about to add it myself.

One thing unnoticed, is that the selection only still doesn't work properly. Small uibdesign issue is see is that options, the booleans, look the same as an operator button now. It's kinda confussing

**Comment by schroef** [comment]
I did noticed it only works only collections, if nothing has material. There are 2 functions called when adding override_material. The add_default one, does a different approach. So i combined this into one method and added a function get_objects. This now also can do sub-collections. Since this method wasnt working yet.

Other issue i still see is single sided meshes. This is due to the node setup and i dont think this can be "fixed".

The black parts is an issue with low transparency settings. For this shot i need to crank it up to 500!
But the the trees and other parts, which are not a closed volume. Will show the cutaway color on the negative normal side.

![particle hair renders emitter visible black2](https://github.com/IfcOpenShell/IfcOpenShell/assets/6923008/5ec339a1-dff1-4d80-aa64-7a008ba84c69)
